mina.netty/src/main/java/org/jboss/netty/channel/socket/nio/AbstractNioWorker.java

```diff
@@ -618,4 +643,53 @@ public void run() {
             }
         });
     }
+
+    public void register(NioChildDatagramChannel childChannel) {    // TODO no datagram specific
+        registerTask(() -> {
+            childChannels.put(childChannel.getId().intValue(), childChannel);
+        });
+    }
+
+    public void deregister(NioChildDatagramChannel childChannel) {
+        registerTask(() -> {
+            childChannels.remove(childChannel.getId().intValue());
+        });
+    }
+
+    public void messageReceived(NioChildDatagramChannel childChannel, Object message) {
+        assert childChannel.getId() >= 0;
+
+        ChannelBuffer buf = (ChannelBuffer) message;
+        ByteBuffer byteBuffer = buf.toByteBuffer();
+        atomicBuffer.wrap(byteBuffer);
+        ringBuffer.write(childChannel.getId(), atomicBuffer, 0, atomicBuffer.capacity());
+    }
+
+    protected void processRead() throws IOException {
+        if (ringBuffer == null) {
+            return;
+        }
+        int workCount = ringBuffer.read(this::handleRead);
+        IdleStrategy idleStrategy = new BackoffIdleStrategy(50, 50, 10000, 20000);
+        idleStrategy.idle(workCount);
+    }
+
+    private void handleRead(int msgTypeId, DirectBuffer buffer, int index, int length) {
+        Channel childChannel = childChannels.get(msgTypeId);
+        if (childChannel == null) {
+            processTaskQueue();
```

Comment:
This also deserves a comment to indicate that the register task may still be in the task queue, so this gives the register task a chance to complete rather than effectively dropping the UDP packet.